lib/bosh/BoshDirectorClient.js

```diff
@@ -16,48 +16,139 @@ const NotFound = errors.NotFound;
 const BadRequest = errors.BadRequest;
 
 class BoshDirectorClient extends HttpClient {
-  constructor(boshDirectorConfig) {
-    const boshDirector = boshDirectorConfig || config.director;
+  constructor() {
     super({
-      baseUrl: boshDirector.url,
-      auth: {
-        user: boshDirector.username,
-        pass: boshDirector.password
-      },
       headers: {
         Accept: 'application/json,text/plain;q=0.9'
       },
-      followRedirect: false,
-      rejectUnauthorized: !boshDirector.skip_ssl_validation
+      followRedirect: false
+    });
+    // activeConfigs - directors supporting lifecycle oprations
+    this.activeConfigs = BoshDirectorClient.getPrimaryConfigs();
+    // activePrimary - directors supporting 'create' oprations
+    this.activePrimary = BoshDirectorClient.getActivePrimary();
+    // oobDirectorConfigs - directors for OutOfBand backup scenario
+    this.oobDirectorConfigs = BoshDirectorClient.getOobDirectorConfigs();
+    this.populateCache();
+  }
+
+  clearCache() {
+    this.cache = {};
+  }
+
+  static getInfrastructure() {
+    return BoshDirectorClient.getActivePrimary()[0].infrastructure;
+  }
+
+  static getPrimaryConfigs() {
+    return _.filter(config.directors, function (director) {
+      return director.primary;
+    });
+  }
+
+  static getActivePrimary() {
+    return _.filter(config.directors, function (director) {
+      return director.primary && director.support_create;
+    });
+  }
+
+  static getOobDirectorConfigs() {
+    return _.filter(config.directors, function (director) {
+      return !director.primary;
+    });
+  }
+
+  getConfigByName(name) {
+    return _.head(_.filter(config.directors, (director) => director.name === name));
+  }
+
+  populateCache() {
+    this.clearCache();
+    return Promise
+      .map(config.directors, directorConfig => {
+        return this
+          .getDeploymentsByConfig(directorConfig)
+          .tap(deployments => _.map(deployments, deployment => this.cache[deployment.name] = directorConfig));
+      })
+      .tap(() => logger.debug('Cached Deployments:', this.cache));
+  }
+
+  deleteCacheEntry(deploymentName) {
+    return delete this.cache[deploymentName];
+  }
+
+  getDirectorConfig(deploymentName) {
+    return Promise.try(() => {
+      logger.info(`Finding the correct director config for:`, deploymentName);
+      const cache_val = this.cache[deploymentName];
+      if (cache_val !== undefined) {
+        logger.debug('found director in cache...', cache_val);
+        return cache_val;
+      }
+      logger.debug('cache miss for', deploymentName);
+      return this
+        .populateCache()
```

Comment:
optimization could be done here , where in instead of rebuidling entire cache we can just get deployment by name and whichever director returns the deployment, we can save its config against the deployment name.